### ISSUE TITLE
Proposal: Provide a Simple Networking Option

### ISSUE THREAD

**Issue by danehans** [contiv/netplugin#106]
Currently, Netplugin requires a modest understanding of the project for users to begin utilizing its benefits. Although Netplugin provides capabilities beyond basic container networking solutions, the barrier of entry may be too high for users/developers that would like to start off with a simple container networking infrastructure and later add complexity.

I propose the following 2 approaches to address this use case:

Option 1. Netplugin provides 2 modes, simple and advanced. Simple mode provides a single tenant, network, etc.. so users can easily spawn containers on a single flat, shared network.

Options 2: Netplugin start with a default config json that provides users with a single flat, shared network to spawn containers.

Whatever approach is developed, it should follow the principle of little-to-no user interaction.

**Comment by erikh** [comment]
I personally like Option #2.

-Erik

**Comment by mapuri** [comment]
yeah, I like '2.' as well for a batteries included model with user not having to worry about specifying labels etc day one to get basic networking going.

Note that netplugin (instance that is run per host) itself doesn't have a user facing interface and is rather supposed to be used in conjunction with an orchestration layer (or a management function). So one way to achieve these defaults (or simple mode configurations) will be to have them setup in the management-function specific thin agent implementations viz. k8s plugin or libnetwork remote-driver.

Would it make sense to explore pushing a default-tenant, network and endpoint configuration in absence of explicit user specified networks labels?

**Comment by danehans** [comment]
@mapuri Regarding the management function, I'm considering Magnum (https://github.com/openstack/magnum) for my use case. Magnum does not use an agent and is essentially a thin wrapper API over existing OpenStack services. Magnum uses Heat (https://wiki.openstack.org/wiki/Heat) for orchestration, which leverages cloud-init and cfn tools: https://github.com/openstack/heat-cfntools

**Comment by jainvipin** [comment]
Option 2 will help if new features are added and default config needs adjustment. In option 1, there is default settings, but it is not explicit. In case user might want to find out, or see what's happening it is not transparent.
+1 for option 2

**Comment by mapuri** [comment]
@danehans thanks for sharing the links.

Below is a way I see a high level workflow that can work with Magnum and Netplugin with option 2. This is mostly based on my quick reading of the Magnum and Heat specs and some assumptions I made around openstack services like neutron, so please correct me wherever I am missing something:
- Netplugin/netmaster is started with a default tenant/network configuration:
   - more tenants and networks can possibly be defined from neutron service (and a neutron agent implementation around netplugin/netmaster).
  - the default configuration oassed to netmaster can be based on a common neutron deployment i.e. vlan or vxlan or L3.
- Magnum continues to create/run pod/container through k8s/swarm API calls:
  - netplugin shall be triggered by using k8s/libnetwork plugin infra as we it integrated today.
  - If magnum passes relevant label information to the pods or containers then netplugin shall use that for provisioning the networks, else the pod/container is placed in default network.

This is very high level, but I am presenting it here to see if this can be a potential integration point and nothing big is missing in my understanding.

There are some changes needed in the way netmaster is run to allow a default configuration being passed. I have filed #108 to track that enhancement.

**Comment by danehans** [comment]
@mapuri I think the workflow you provided is accurate. For Netplugin to use VLAN's for it's default configuration, I believe the VLAN-aware VM's Neutron spec will need to be implemented:

https://review.openstack.org/#/c/94612/21/specs/liberty/vlan-aware-vms.rst

Since VXLAN is becoming the preferred overlay within OpenStack, I suggest the Netplugin default configuration use VXLAN. By using VXLAN, Neutron should only see IP packets with src/dest IP's of the Nova instances that run containers. This should make the container networking overlay transparent to Neutron. When the VLAN-aware VM's spec is implemented, I could see a need for coordinating VLAN assignment between Neutron and Netplugin so the same VLAN is stitched across the physical/vm/container path.

Magnum should be capable of optionally passing labels through k8s/swarm API calls. I will need to update the Magnum templates to optionally allow an array of labels to be passed to Heat for orchestration purposes.
